### ISSUE TITLE
Learning to Rank with --csoaa_ldf

### ISSUE THREAD

**Issue by usptact** [JohnLangford/vowpal_wabbit#1354]
Hi,

I am trying to build a learning to rank (l2r) model. It looks like context sensitive one-against-all with label dependent features options (csoaa_ldf) is the feature of VW to use.

I downloaded the MQ2007 dataset and converted it into VW format (multiline format). Each block of lines is one query matched to a bunch of documents. The rank label (cost 0 - lowest rank to cost 2 - highest rank) was converted to costs (opposite order): label 1 - lowest rank to label 3 - highest rank.

If an example was of lowest rank, the entry looks like this:
`1:0 2:1 3:2 |f ....
`

If an example was of highest rank, the entry looks like this:
`1:2 2:1 3:0 |f ...
`

The middle rank example looks like this:
`1:1 2:0 3:1 |f ...
`

I ran training on a dataset (shuffled blocks) like this:
`vw -d train.vw --csoaa_ldf=m --passes 5 -c -k -f mq2007.model --l2 0.000001
`

The prediction of ranks was performed on similarly formated dataset. Each example got all possible labels w/o costs (must tell which labels to predict).
`vw -t -d test.vw -i mq2007.model -r test.predictions
`

The problem is that I constantly get the label 1 all weight and labels 2 and 3 get always 0. And this even on the training set. I tried csoaa_ldf=mc, using more passes, setting --l2 to different values to no avail. I tried also 0/1 costs to simulate relevant/non-relevant cases.

Is there something that I overlooked or prepared data incorrectly?

EDIT: After carefully checking data format, I noticed that the label distributions in training set (globally, disregarding blocks) are fairly imbalanced:
1 - 41750 feature vectors
2 - 11273 feature vectors
3 - 2948 feature vectors

For this dataset it is expected: there is only one or very few relevant documents per query, more middle rank and many of lowest rank.

Can this be the reason why label 1 is always predicted?

The data (in VW format) can be downloaded from OneDrive:
https://1drv.ms/u/s!AuktSSr-TPh5iRs8-2kGM5o1wfBZ

I compiled the latest VW (ver. 8.4.0) from 11/14/2017 from GitHub.

Thanks

**Comment by JohnLangford** [comment]
@hal3 may be able to comment further, but I believe you are inputting the data in the wrong format for ldf.  In particular, you want:

<label_1>:<cost_1> | features of label 1
<label_2>:<cost_2> | features of label 2
...

**Comment by JohnLangford** [comment]
See here: https://github.com/JohnLangford/vowpal_wabbit/blob/master/test/train-sets/cs_test.ldf

**Comment by usptact** [comment]
@JohnLangford Thanks!

I think that I am already following the format.

The first integer is a label, the second is the cost. I use the costs to indicate feature vectors that are of highest (0 cost), middle (1 cost) and lowest (2 cost) ranks.

For example:
1:1 2:0 3:1 |f a b c d
1:0 2:1 3:2 |f e f g h
1:2 2:1 3:0 |f x y z

The first vector is of of rank 2 (label+cost is 2:0). The second vector is of rank 1 (label+cost is 1:0). Finally, the third vector is of rank 3 (label+cost is 3:0). In each vector, the other classes have non-zero costs to indicate the order.

Am I still missing something?

**Comment by usptact** [comment]
I changed the format according to the example but I still see the same bad predictions. The trace of scores during training shows suspiciously very high error rates. Something is wrong here.

**Comment by usptact** [comment]
```
vw -d train.vw --csoaa_ldf=m --passes 5 -c -k -f mq2007.model --l2 0.0000001
using l2 regularization = 1e-07
final_regressor = mq2007.model
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
decay_learning_rate = 1
creating cache_file = train.vw.cache
Reading datafile = train.vw
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
2.000000 2.000000            1            1.0    known        1     1190
1.000000 0.000000            2            2.0    known        0      980
1.750000 2.500000            4            4.0    known        1     1216
1.625000 1.500000            8            8.0    known        0     1520
1.937500 2.250000           16           16.0    known        1     1114
1.687500 1.437500           32           32.0    known        0      962
1.687500 1.687500           64           64.0    known        0     1305
1.695312 1.703125          128          128.0    known        0     1250
1.613281 1.531250          256          256.0    known        0      985
1.542969 1.472656          512          512.0    known        0     1148
1.509766 1.476562         1024         1024.0    known        0     1215
 unknown  unknown         2048         2048.0    known        0      956 h
 unknown  unknown         4096         4096.0    known        0     1094 h

finished run
number of examples per pass = 1356
passes used = 4
weighted example sum = 5424.000000
weighted label sum = 0.000000
average loss = undefined (no holdout)
total feature number = 6035752
```

**Comment by hal3** [comment]
Can you update the dataset so I can download and take a look?

**Comment by usptact** [comment]
@hal3 Thanks for looking into this! The updated dataset is uploaded:
https://1drv.ms/u/s!AuktSSr-TPh5iRxs_0cGTuptO69s

I made a change how the data labels and costs are converted from the original SVM-Light format. I noticed that examples can have different number of ranks. Before I blindly converted a label from original file to VW format and attached a cost according to the label. This resulted in some examples having no feature vectors with cost 0 (the top ranked).

Take a look at the "sample.txt" (original) and "sample.vw" (converted) how I made prepared labels and costs.

In nutshell, I take an example and find how many labels (ranks) are there. I subtract from this number the current label + 1 to get the VW cost. This ensures that top label in original file maps to 0 cost. 0 label gets mapped to the highest cost.

**Comment by usptact** [comment]
```
$ vw -d train.vw --csoaa_ldf=m --passes 5 -c -k -f mq2007.model --l2 0.0000001
using l2 regularization = 1e-07
final_regressor = mq2007.model
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
decay_learning_rate = 1
creating cache_file = train.vw.cache
Reading datafile = train.vw
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.000000 0.000000            1            1.0    known        1      948
0.500000 1.000000            2            2.0    known        1     1180
0.250000 0.000000            4            4.0    known        0      900
0.500000 0.750000            8            8.0    known        0     1097
0.812500 1.125000           16           16.0    known        0     1056
0.906250 1.000000           32           32.0    known        1     1248
0.734375 0.562500           64           64.0    known        0     1116
0.742188 0.750000          128          128.0    known        1     1177
0.726562 0.710938          256          256.0    known        0     1094
0.728516 0.730469          512          512.0    known        0      911
0.785156 0.841797         1024         1024.0    known        0     1520
 unknown  unknown         2048         2048.0    known        1     1342 h
 unknown  unknown         4096         4096.0    known        0      967 h

finished run
number of examples per pass = 1356
passes used = 4
weighted example sum = 5424.000000
weighted label sum = 0.000000
average loss = undefined (no holdout)
total feature number = 6035752
```

**Comment by JohnLangford** [comment]
You seem to be assigning some meaning to the label id here, but that's not how the LDF format works.  In essence, the label is purely defined by the set of features given to the label. 

You also refer to 0 as high cost, but lower costs have lower value intrinsically.

**Comment by JohnLangford** [comment]
Hope this help---I'm closing for now, but reopen as necessary.
